nntrainer/graph/network_graph.cpp

```diff
@@ -0,0 +1,424 @@
+  std::string orig_name = layer->getName();
+  bool orig_name_empty = orig_name.empty();
+  if (!orig_name_empty && !force_rename &&
+      layer_names.end() == layer_names.find(orig_name))
+    return;
+
+  /** If just prefix with layer name makes it unique - directly set the name */
+  if (!orig_name_empty) {
+    std::string direct_name = prefix + orig_name;
+    if (layer_names.find(direct_name) == layer_names.end()) {
+      layer->setName(direct_name);
+      return;
+    }
+  }
+
+  std::set<std::string>::iterator iter;
+  std::string name;
+  if (orig_name_empty)
+    orig_name = layer->getType();
+  std::string direct_name = prefix + orig_name;
+
+  do {
+    name = direct_name + std::to_string(def_name_count++);
+    iter = layer_names.find(name);
+  } while (iter != layer_names.end());
+
+  layer->setName(name);
+}
+
+int NetworkGraph::realizeMultiInputType(Layer &current) {
+  int status = ML_ERROR_NONE;
+  if (current.num_inputs == 1)
+    return ML_ERROR_NONE;
+
+  if (current.num_inputs > 1) {
+    std::shared_ptr<Layer> layer = nntrainer::createLayer("concat");
+    ensureName(layer, current.getName());
+    layer->num_inputs = current.num_inputs;
+    layer->input_layers.clear();
+    for (unsigned int i = 0; i < current.input_layers.size(); ++i)
+      layer->input_layers.push_back(current.input_layers[i]);
+
+    current.num_inputs = 1;
+    current.input_layers.clear();
+    current.input_layers.push_back(layer->getName());
+    addLayerNode(layer);
+  }
+  return status;
+}
+
+int NetworkGraph::realizeFlattenType(
+  Layer &current, std::vector<std::shared_ptr<Layer>> layers) {
+  if (num_node == 0) {
+    ml_loge("layer is empty");
+    return ML_ERROR_INVALID_PARAMETER;
+  }
+
+  if (current.getType() == "flatten") {
+    ml_loge(
+      "It is not allowed to realize flatten layer, possibly flatten layer is "
+      "added right after flatten");
+    return ML_ERROR_INVALID_PARAMETER;
+  }
+
+  std::shared_ptr<Layer> layer = nntrainer::createLayer("flatten");
+
+  ensureName(layer, current.getName());
+
+  layer->num_inputs = 1;
+  layer->input_layers.clear();
+  layer->input_layers.push_back(current.getName());
+
+  addLayerNode(layer);
+
+  for (unsigned int i = 0; i < layers.size(); ++i) {
+    for (unsigned int j = 0; j < layers[i]->input_layers.size(); ++j) {
+      if (layers[i]->input_layers[j] == current.getName()) {
+        layers[i]->input_layers[j] = layer->getName();
+      }
+    }
+  }
+
+  return ML_ERROR_NONE;
+}
+
+int NetworkGraph::realizeActivationType(
+  Layer &current, std::vector<std::shared_ptr<Layer>> layers) {
+  ActivationType act = current.getActivationType();
+
+  if (act == ActivationType::ACT_NONE) {
+    /// ActivationType::ACT_NONE does not need realization
+    return ML_ERROR_NONE;
+  }
+
+  if (num_node == 0) {
+    ml_loge("layer is empty");
+    return ML_ERROR_INVALID_PARAMETER;
+  }
+
+  if (current.getType() == "activation") {
```

Comment:
if (current.getType() == ActivationLayer::type) {

Reply:
done

Reply:
@zhoonit In this case, as we use getType() which is layers internal function, and compare it against its own type, is istrequal required? When comparing externally provided string, its understandable, but when working with internal strings, given its allocated by the framework itself, I think direct should also work.

Reply:
We know this will work for internal usage, however semantic states type is case insensitive, if the part we assumed it would stay internal happens to have external usage, it will be very hard to catch the error. 
Just my opinion though, If we start to tell internal and external types, we might be ending up into a situation hard to deal with. 

If there is no visible performance improvement, I think it is better to stick to semantics.
If we have a performance problem, I think we'd better change semantics (this is not easy though) or at least strictly restrict the usage.